File: apps/windows/TeslaSync.App/modals-dialogs/AcknowledgeAlertDialog.Model.cs
using System.Globalization;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.ModalsDialogs;

/// <summary>
/// The acknowledgement draft the dialog emits — the native analogue of the web <c>onSubmit(note)</c> argument
/// (web/src/features/admin/components/AcknowledgeAlertDialog.tsx). <see cref="Note"/> is the trimmed note, which
/// may be the empty string when the user leaves the textarea blank: the backend treats an empty/whitespace note
/// as "ack with no note" so the audit timeline still records who + when (web handleSubmit comment).
/// </summary>
public sealed record AcknowledgeAlertDraft(string Note);

/// <summary>
/// Canonical bounds, the note length policy and i18n keys for the <c>AcknowledgeAlertDialog</c> surface — the
/// native mirror of <c>web/src/features/admin/components/AcknowledgeAlertDialog.tsx</c>. The web component ships
/// literal copy behind <c>react-i18next</c> keys; every literal is keyed here (with that literal as the English
/// fallback) so the native view and view-model stay free of inline strings and resolve through the i18n facade.
/// UI-free so every key, fallback and bound is asserted headlessly.
/// </summary>
public static class AcknowledgeAlertRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "AcknowledgeAlertDialog";

    /// <summary>Maximum trimmed note length before the field is flagged too long (web <c>NOTE_MAX</c> = 1000).</summary>
    public const int NoteMaxLength = 1000;

    /// <summary>
    /// Hard cap on the editable text length (web <c>maxLength={NOTE_MAX + 50}</c> = 1050). The 50-char slack lets
    /// the user type a little past the limit so the too-long validation error can surface rather than silently
    /// truncating at the boundary.
    /// </summary>
    public const int NoteInputMaxLength = NoteMaxLength + 50;

    /// <summary>Dialog title (web <c>t('alerts.ack.dialogTitle', 'Acknowledge alert')</c>).</summary>
    public static string DialogTitle(ILocalizer localizer) =>
        Require(localizer).GetString("alerts.ack.dialogTitle", "Acknowledge alert");

    /// <summary>Note field label (web <c>t('alerts.ack.noteLabel', 'Note (optional)')</c>).</summary>
    public static string NoteLabel(ILocalizer localizer) =>
        Require(localizer).GetString("alerts.ack.noteLabel", "Note (optional)");

    /// <summary>Note field input hint — the web note field's hint text "Optional: what's being done?".</summary>
    public static string NotePrompt(ILocalizer localizer) =>
        Require(localizer).GetString("alerts.ack.notePlaceholder", "Optional: what's being done?"); // parity:allow web i18n key (input hint text), not a stub

    /// <summary>Cancel button label (web <c>t('alerts.ack.cancel', 'Cancel')</c>).</summary>
    public static string CancelLabel(ILocalizer localizer) =>
        Require(localizer).GetString("alerts.ack.cancel", "Cancel");

    /// <summary>Acknowledge (submit) button label (web <c>t('alerts.ack.submit', 'Acknowledge')</c>).</summary>
    public static string SubmitLabel(ILocalizer localizer) =>
        Require(localizer).GetString("alerts.ack.submit", "Acknowledge");

    /// <summary>
    /// The note hint with the character cap interpolated — the native analogue of
    /// <c>t('alerts.ack.noteHint', 'Up to {{max}} characters. Shared in the audit timeline.', { max: NOTE_MAX })</c>.
    /// Mirrors the FeedbackModal interpolation convention: resolve the keyed template then substitute the
    /// i18next <c>{{max}}</c> token so a translated catalog string keeps the substitution point.
    /// </summary>
    public static string NoteHint(ILocalizer localizer, int max)
    {
        string template = Require(localizer)
            .GetString("alerts.ack.noteHint", "Up to {{max}} characters. Shared in the audit timeline.");
        return template.Replace("{{max}}", max.ToString(CultureInfo.CurrentCulture), StringComparison.Ordinal);
    }

    private static ILocalizer Require(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer;
    }
}

/// <summary>
/// Pure projections for the <c>AcknowledgeAlertDialog</c> surface — the native analogue of the web component's
/// <c>note.trim()</c> + <c>trimmed.length > NOTE_MAX</c> logic and the <c>onSubmit(trimmed)</c> argument
/// assembly. UI-free so the note-length policy and submit-gate are unit-tested headlessly and the view-model
/// never recomputes them inline.
/// </summary>
public static class AcknowledgeAlertProjection
{
    /// <summary>The trimmed note (web <c>note.trim()</c>); never null.</summary>
    public static string NormalizeNote(string? note) => (note ?? string.Empty).Trim();

    /// <summary>
    /// True once the trimmed note exceeds <see cref="AcknowledgeAlertRegistration.NoteMaxLength"/>
    /// (web <c>trimmed.length > NOTE_MAX</c>): the submit gate closes and the field shows the too-long error.
    /// </summary>
    public static bool IsTooLong(string? note) =>
        NormalizeNote(note).Length > AcknowledgeAlertRegistration.NoteMaxLength;

    /// <summary>
    /// Assemble the <c>onSubmit</c> argument from the current note — the native analogue of
    /// <c>onSubmit(note.trim())</c>. The note is trimmed (and may be the empty string, which the backend accepts
    /// as an ack with no note).
    /// </summary>
    public static AcknowledgeAlertDraft BuildDraft(string? note) => new(NormalizeNote(note));
}

/// <summary>
/// PII-safe diagnostics for the <c>AcknowledgeAlertDialog</c> surface (P1/S11 diagnostics contract). Records only
/// operational counters with the surface slug — never the acknowledgement note or the acked alert's title — so a
/// diagnostics line can never leak the audit-note content. Thread-safe.
/// </summary>
public sealed class AcknowledgeAlertDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;
    private long _acknowledged;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public AcknowledgeAlertDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Number of acknowledgements emitted from this surface.</summary>
    public long Acknowledged => Interlocked.Read(ref _acknowledged);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=AcknowledgeAlertDialog</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke(string.Create(
            CultureInfo.InvariantCulture, $"view.opened slug={AcknowledgeAlertRegistration.Slug}"));
    }

    /// <summary>Record that an acknowledgement was emitted (the note / alert title are never logged).</summary>
    public void RecordAcknowledged()
    {
        Interlocked.Increment(ref _acknowledged);
        _sink?.Invoke(string.Create(
            CultureInfo.InvariantCulture, $"alert.acknowledged slug={AcknowledgeAlertRegistration.Slug}"));
    }
}

File: apps/android/app/src/main/java/com/teslasync/modals-dialogs/AcknowledgeAlertDialog/AcknowledgeAlertDialogModel.kt
// Pure, framework-free model + projection for the AcknowledgeAlertDialog modal/dialog surface — the native analogue of
// everything the web component derives before it returns JSX (web/src/features/admin/components/AcknowledgeAlertDialog.tsx).
// No Compose, no Android, no HTTP: every declaration here is exercised off-device by the :android:testReleaseUnitTest
// gate, so the composable stays a thin render layer over these pure functions.
//
// The web component is the modal opened from an alert row's "Acknowledge" button. It is a *controlled* form — its only
// hooks are `useTranslation` (i18n, P1/S10) and `useId` (the hint's a11y association); it performs no data fetch and
// owns no store. It records an optional free-text note (the input hard-caps at `NOTE_MAX + 50` characters, web
// `maxLength`, while the trimmed note must stay within `NOTE_MAX` or the field flips to its error state and the submit
// disables) and hands the *trimmed* note back to the parent's `onSubmit` callback. An empty / whitespace-only note is
// accepted — the backend treats it as "ack with no note" so the audit timeline still captures who + when. This file
// owns the data derivations behind that form: the maxLength clamp (web `maxLength={NOTE_MAX + 50}`), the trim (web
// `note.trim()`), the over-limit guard (web `tooLong`), the submit-enablement guard (web `submitting || tooLong`), and
// the trimmed-note resolution handed to `onSubmit` (web `onSubmit(trimmed)`). The localized labels are resolved at the
// Compose boundary, never here.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/modals-dialogs/AcknowledgeAlertDialog — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package intentionally
// diverges from the path — exactly as the sibling modal/dialog surfaces do. `MatchingDeclarationName` is suppressed for
// the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.modalsdialogs.acknowledgealertdialog

import io.teslasync.shared.core.diagnostics.Logger

/**
 * The pure derivations the composable renders over — the native mirror of the web component's inline note state, the
 * derived `trimmed` / `tooLong` values, and the `handleSubmit` guard. Stateless and side-effect-free, so it is fully
 * covered by the off-device unit gate.
 */
object AcknowledgeAlertProjection {
    /** Maximum trimmed note length the client accepts before the field flips to its error state (web `NOTE_MAX`). */
    const val NOTE_MAX: Int = 1000

    /**
     * Extra typing headroom the textarea allows past [NOTE_MAX] so a user can over-type and trim back without the
     * input swallowing keystrokes (web `maxLength={NOTE_MAX + 50}`).
     */
    const val NOTE_INPUT_GRACE: Int = 50

    /** The textarea's hard character cap — the web `maxLength` (`NOTE_MAX + 50`). */
    const val MAX_INPUT_LENGTH: Int = NOTE_MAX + NOTE_INPUT_GRACE

    /** Clamps a note edit to the textarea's accepted maximum (web `maxLength={NOTE_MAX + 50}`). */
    fun clampNote(note: String): String = note.take(MAX_INPUT_LENGTH)

    /** The trimmed note the dialog submits and length-checks (web `const trimmed = note.trim()`). */
    fun trimmedNote(note: String): String = note.trim()

    /**
     * Whether the trimmed note exceeds [NOTE_MAX] — the web `tooLong`. Drives both the field's error affordance and the
     * disabled submit. An empty / whitespace-only note is never too long (web parity: blank notes are allowed).
     */
    fun isTooLong(note: String): Boolean = trimmedNote(note).length > NOTE_MAX

    /**
     * Whether the Acknowledge action may fire — the web `handleSubmit` guard `if (submitting || tooLong) return` (and
     * the button's `disabled={submitting || tooLong}`). Note an empty note is submittable, exactly as the web allows.
     */
    fun canSubmit(
        note: String,
        submitting: Boolean,
    ): Boolean = !submitting && !isTooLong(note)

    /** The value handed to the parent's `onSubmit` — the trimmed note, which may be the empty string (web `trimmed`). */
    fun resolveSubmitNote(note: String): String = trimmedNote(note)
}

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object AcknowledgeAlertDialogRegistration {
    /** Stable surface id. */
    const val ID: String = "acknowledge-alert-dialog"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "AcknowledgeAlertDialog"
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface
 * [AcknowledgeAlertDialogRegistration.SLUG] — never the typed note or the acknowledged alert's title — so a diagnostics
 * line can never leak what the operator is acknowledging or writing.
 */
object AcknowledgeAlertDialogDiagnostics {
    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to AcknowledgeAlertDialogRegistration.SLUG))
    }
}
